In [14]:
import pandas as pd
sp = pd.read_csv('StudentsPerformance.csv')
sp

gender race/ethnicity parental level of education         lunch  \
0    female        group C            some high school  free/reduced   
1    female        group B                 high school  free/reduced   
2    female        group B            some high school  free/reduced   
3    female        group B                some college      standard   
4    female        group C                some college  free/reduced   
..      ...            ...                         ...           ...   
995    male        group E                some college      standard   
996    male        group A                some college      standard   
997    male        group D                some college      standard   
998    male        group E          associate's degree  free/reduced   
999    male        group E           bachelor's degree      standard   

    test preparation course  math score  reading score  writing score  
0                      none           0             17             10  
1                      none           8             24             23  
2                      none          18             32             28  
3                      none          11             38             32  
4                      none          22             39             33  
..                      ...         ...            ...            ...  
995               completed          99             87             81  
996               completed         100             96             86  
997               completed         100             97             99  
998               completed         100            100             93  
999               completed         100            100            100  

[1000 rows x 8 columns]

In [3]:
sp.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 8 columns):
 #   Column                       Non-Null Count  Dtype 
---  ------                       --------------  ----- 
 0   gender                       1000 non-null   object
 1   race/ethnicity               1000 non-null   object
 2   parental level of education  1000 non-null   object
 3   lunch                        1000 non-null   object
 4   test preparation course      1000 non-null   object
 5   math score                   1000 non-null   int64 
 6   reading score                1000 non-null   int64 
 7   writing score                1000 non-null   int64 
dtypes: int64(3), object(5)
memory usage: 62.6+ KB


# 1- Number of males and females participated in the test

In [4]:
sp['gender'].value_counts()

female    518
male      482
Name: gender, dtype: int64

### There are 518 female and 482 male students in the given data

# 2- Parental level of education

In [5]:
sp['parental level of education'].value_counts()

some college          226
associate's degree    222
high school           196
some high school      179
bachelor's degree     118
master's degree        59
Name: parental level of education, dtype: int64

### Most of the parents have a good education qualification and 59 of them have a master's degree.Few of them have only high school education.

# 3(a)- Most on average for math, reading and writing based on gender

#### - Most on average for math

In [4]:
df1 = sp.groupby('gender')['math score'].mean()
df1

gender
female    63.376448
male      68.821577
Name: math score, dtype: float64

#### - Most on average for reading

In [8]:
df2 = sp.groupby('gender')['reading score'].mean()
df2

gender
female    72.590734
male      65.545643
Name: reading score, dtype: float64

#### - Most on average for writing

In [9]:
df3 = sp.groupby('gender')['writing score'].mean()
df3

gender
female    72.467181
male      63.446058
Name: writing score, dtype: float64

### Female students are scored high average in reading and writing ,but in maths male students have the high average score

# 3(b)- Most on average for math, reading and writing based on  Test preparation course

#### - Most on average for math

In [6]:
df4 = sp.groupby('test preparation course')['math score'].mean()
df4

test preparation course
completed    69.96648
none         63.78972
Name: math score, dtype: float64

#### - Most on average for reading

In [5]:
df5 = sp.groupby('test preparation course')['reading score'].mean()
df5

test preparation course
completed    74.175978
none         66.417445
Name: reading score, dtype: float64

#### - Most on average for writing

In [7]:
df6 = sp.groupby('test preparation course')['writing score'].mean()
df6

test preparation course
completed    74.684358
none         64.457944
Name: writing score, dtype: float64

### Students who completed test preparation course are scored high average in maths, reading and writing.

# 4- Scoring Variation Based on

## a)Gender
   #### i) Math score

In [31]:
sp_var = sp
sp_var['math var'] = (sp3['math score']-sp3['math score'].mean())**2
sp_var['reading var'] = (sp_var['reading score']-sp_var['reading score'].mean())**2
sp_var['writing var'] = (sp_var['writing score']-sp_var['writing score'].mean())**2
sp_var

gender race/ethnicity parental level of education         lunch  \
0    female        group C            some high school  free/reduced   
1    female        group B                 high school  free/reduced   
2    female        group B            some high school  free/reduced   
3    female        group B                some college      standard   
4    female        group C                some college  free/reduced   
..      ...            ...                         ...           ...   
995    male        group E                some college      standard   
996    male        group A                some college      standard   
997    male        group D                some college      standard   
998    male        group E          associate's degree  free/reduced   
999    male        group E           bachelor's degree      standard   

    test preparation course  math score  reading score  writing score  \
0                      none           0             17             10   
1                      none           8             24             23   
2                      none          18             32             28   
3                      none          11             38             32   
4                      none          22             39             33   
..                      ...         ...            ...            ...   
995               completed          99             87             81   
996               completed         100             96             86   
997               completed         100             97             99   
998               completed         100            100             93   
999               completed         100            100            100   

        math var  reading var  writing var  
0    4356.132001  2724.318025  3377.818161  
1    3364.116001  2042.588025  2035.724161  
2    2304.096001  1383.468025  1609.534161  
3    3025.110001   973.128025  1304.582161  
4    1936.088001   911.738025  1233.344161  
..           ...          ...          ...  
995  1088.934001   317.018025   165.920161  
996  1155.932001   718.508025   319.730161  
997  1155.932001   773.118025   953.636161  
998  1155.932001   948.948025   619.064161  
999  1155.932001   948.948025  1016.398161  

[1000 rows x 11 columns]

In [32]:
sp3.groupby('gender')['math var'].sum()/sp3['gender'].value_counts()

gender
female    263.350808
male      219.405146
dtype: float64

#### ii) Reading score

In [33]:
sp3.groupby('gender')['reading var'].sum()/sp3['gender'].value_counts()

gender
female    218.807523
male      213.113440
dtype: float64

#### iii) Writing score

In [34]:
sp3.groupby('gender')['writing var'].sum()/sp3['gender'].value_counts()

gender
female    238.850586
male      223.830364
dtype: float64

### In all subjects, scoring variation of female students are high when compared with scoring variation of male students 

## b)Test Preparation Course
#### i) Math score

In [35]:
sp3.groupby('test preparation course')['math var'].sum()/sp3['test preparation course'].value_counts()

completed    226.020001
none         251.174203
dtype: float64

#### ii) Reading score

In [36]:
sp3.groupby('test preparation course')['reading var'].sum()/sp3['test preparation course'].value_counts()

completed    207.564086
none         220.802231
dtype: float64

#### iii) Writing score

In [37]:
sp3.groupby('test preparation course')['writing var'].sum()/sp3['test preparation course'].value_counts()

completed    217.817139
none         239.302653
dtype: float64

### In all subjects, scoring variation of students who completed test preparation course are less when compared with scoring variation of students who doesn't complete test preparation course

# 5- Top 25% students to give bonus points

In [51]:
total = sp.sort_values('math score',ascending=True)
total['math score'].describe()

count    1000.000000
mean       66.001000
std        15.569567
min         0.000000
25%        57.000000
50%        66.000000
75%        77.000000
max       100.000000
Name: math score, dtype: float64

### Here we ordered the given data in ascending order based on their maths score and found the 3rd quartile of maths score. ie; 77 marks(75% students scored below 77 marks and 25% students scored 77 marks or above).That is the students who scored 77 marks or above are eligible for the bonus points. 

In [53]:
sp[sp['math score']>=77]

gender race/ethnicity parental level of education         lunch  \
414  female        group A                 high school  free/reduced   
415  female        group B             master's degree  free/reduced   
416  female        group B           bachelor's degree  free/reduced   
417  female        group B             master's degree      standard   
418  female        group B                 high school      standard   
..      ...            ...                         ...           ...   
995    male        group E                some college      standard   
996    male        group A                some college      standard   
997    male        group D                some college      standard   
998    male        group E          associate's degree  free/reduced   
999    male        group E           bachelor's degree      standard   

    test preparation course  math score  reading score  writing score  
414               completed          77             88             85  
415               completed          77             97             94  
416                    none          77             85             87  
417                    none          77             90             84  
418               completed          77             82             89  
..                      ...         ...            ...            ...  
995               completed          99             87             81  
996               completed         100             96             86  
997               completed         100             97             99  
998               completed         100            100             93  
999               completed         100            100            100  

[255 rows x 8 columns]

#### There are 255 students who scored 77 marks and above,that is 25.5% of total students.
#### If we do a sorting of top 25% students based on their math score as shown below, according to the above data 5 eligible students will lose their bonus point.

In [49]:
sp.sort_values('math score',ascending=False).head(250)

gender race/ethnicity parental level of education         lunch  \
999    male        group E           bachelor's degree      standard   
996    male        group A                some college      standard   
515  female        group E                some college      standard   
517  female        group E          associate's degree      standard   
516  female        group E           bachelor's degree      standard   
..      ...            ...                         ...           ...   
856    male        group E            some high school      standard   
855    male        group E          associate's degree  free/reduced   
854    male        group D            some high school      standard   
853    male        group D          associate's degree  free/reduced   
852    male        group D                some college  free/reduced   

    test preparation course  math score  reading score  writing score  
999               completed         100            100            100  
996               completed         100             96             86  
515                    none         100             92             97  
517                    none         100            100            100  
516                    none         100            100            100  
..                      ...         ...            ...            ...  
856               completed          77             76             77  
855               completed          77             69             68  
854               completed          77             68             69  
853                    none          77             78             73  
852                    none          77             62             64  

[250 rows x 8 columns]